clash-cores/src/Clash/Cores/Sgmii/Sync.hs

```diff
@@ -0,0 +1,284 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module Clash.Cores.Sgmii.Sync where
+
+import Clash.Cores.Sgmii.Common
+import Clash.Cores.Sgmii.EbTb
+import Clash.Prelude
+
+-- | State type of 'sync'. This contains all states as they are defined in IEEE
+--   802.3 Clause 36.
+data SyncState
+  = LossOfSync {_rd :: Bool, _rxEven :: RxEven}
+  | CommaDetect1 {_rd :: Bool, _rxEven :: RxEven}
+  | AcquireSync1 {_rd :: Bool, _rxEven :: RxEven}
+  | CommaDetect2 {_rd :: Bool, _rxEven :: RxEven}
+  | AcquireSync2 {_rd :: Bool, _rxEven :: RxEven}
+  | CommaDetect3 {_rd :: Bool, _rxEven :: RxEven}
+  | SyncAcquired1 {_rd :: Bool, _rxEven :: RxEven}
+  | SyncAcquired2 {_rd :: Bool, _rxEven :: RxEven}
+  | SyncAcquired2A {_rd :: Bool, _rxEven :: RxEven, _goodCgs :: Index 4}
+  | SyncAcquired3 {_rd :: Bool, _rxEven :: RxEven}
+  | SyncAcquired3A {_rd :: Bool, _rxEven :: RxEven, _goodCgs :: Index 4}
+  | SyncAcquired4 {_rd :: Bool, _rxEven :: RxEven}
+  | SyncAcquired4A {_rd :: Bool, _rxEven :: RxEven, _goodCgs :: Index 4}
```

Comment:
Are these actually different states? It seems like you are using the constructors as counter.

Reply:
There are states as defined in IEEE 802.3 Clause 36, Figure 36-9 to be specific.